// astar-typescript/core/Node.js
"use strict";
/**
* @author       Eric Kuhn <digit.sensitivee@gmail.com>
* @copyright    2017 Eric Kuhn
* @license      {@link https://opensource.org/licenses/MIT|MIT License}
*/
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The node class.
 * Each node has general properties like the id, the x and y position.
 * There are other more specific properties.
 * It is possible to define other properties as needed.
 */
class Node {
    /* Getter functions */
    getId() { return this.m_id; }
    getPositionX() { return this.m_posX; }
    getPositionY() { return this.m_posY; }
    getFValue() { return this.m_fValue; }
    getGValue() { return this.m_gValue; }
    getHValue() { return this.m_hValue; }
    getParent() { return this.m_parent; }
    getIsOnClosedList() { return this.m_isOnClosedList; }
    getIsOnOpenList() { return this.m_isOnOpenList; }
    getIsWalkable() { return this.m_isWalkable; }
    /* Setter functions */
    setFValue() { this.m_fValue = this.m_gValue + this.m_hValue; }
    setGValue(_gValue) { this.m_gValue = _gValue; }
    setHValue(_hValue) { this.m_hValue = _hValue; }
    setParent(_parent) { this.m_parent = _parent; }
    setIsOnClosedList(_isOnClosedList) { this.m_isOnClosedList = _isOnClosedList; }
    setIsOnOpenList(_isOnOpenList) { this.m_isOnOpenList = _isOnOpenList; }
    setIsWalkable(_isWalkable) { this.m_isWalkable = _isWalkable; }
    constructor(aParams) {
        /* Set the general properties */
        this.m_id = aParams.id;
        this.m_posX = aParams.xPos;
        this.m_posY = aParams.yPos;
        /* Set the specific properties */
        this.m_hValue = 0;
        this.m_gValue = 0;
        this.m_fValue = 0;
        this.m_parent = undefined;
        this.m_isOnClosedList = false;
        this.m_isOnOpenList = false;
        this.m_isWalkable = aParams.walkable || true;
    }
}
exports.Node = Node;
